File: Data_Structures/Queue/CircularQueue/JavaScript/CircularQueue.js
/*
 * CircularQueue - Fixed-capacity circular array queue
 *
 * Operations: enqueue O(1), dequeue O(1), getFront/getRear O(1)
 */

class CircularQueue {
    constructor(capacity) {
        this._arr   = new Array(capacity).fill(null);
        this._cap   = capacity;
        this._front = 0;
        this._rear  = -1;
        this._count = 0;
    }

    isEmpty() { return this._count === 0; }
    isFull()  { return this._count === this._cap; }
    size()    { return this._count; }

    enqueue(value) {
        if (this.isFull()) throw new Error("Queue is full");
        this._rear = (this._rear + 1) % this._cap;
        this._arr[this._rear] = value;
        this._count++;
    }

    dequeue() {
        if (this.isEmpty()) throw new Error("Queue is empty");
        const val = this._arr[this._front];
        this._arr[this._front] = null;
        this._front = (this._front + 1) % this._cap;
        this._count--;
        return val;
    }

    getFront() {
        if (this.isEmpty()) throw new Error("Queue is empty");
        return this._arr[this._front];
    }

    getRear() {
        if (this.isEmpty()) throw new Error("Queue is empty");
        return this._arr[this._rear];
    }
}

const cq = new CircularQueue(5);
cq.enqueue(10);
cq.enqueue(20);
cq.enqueue(30);
cq.enqueue(40);
cq.enqueue(50);

console.log("Front:", cq.getFront(), "Rear:", cq.getRear());
console.log("Dequeue:", cq.dequeue());
console.log("Dequeue:", cq.dequeue());

// Wrap-around test
cq.enqueue(60);
cq.enqueue(70);
console.log("After wrap — Front:", cq.getFront(), "Rear:", cq.getRear());

File: Data_Structures/Queue/CircularQueue/Python/CircularQueue.py
"""
CircularQueue - fixed-capacity circular queue using list

Operations:
- enqueue(value) : raises OverflowError if full
- dequeue() : raises IndexError if empty
- get_front(), get_rear(), is_empty(), is_full(), size(), clear()

Complexity: all operations O(1)
"""


class CircularQueue:
    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError("capacity must be > 0")
        self._arr = [None] * capacity
        self._cap = capacity
        self._front = 0
        self._rear = -1
        self._count = 0

    def enqueue(self, value):
        if self.is_full():
            raise OverflowError("Queue is full")
        self._rear = (self._rear + 1) % self._cap
        self._arr[self._rear] = value
        self._count += 1

    def dequeue(self):
        if self.is_empty():
            raise IndexError("dequeue from empty queue")
        val = self._arr[self._front]
        self._arr[self._front] = None
        self._front = (self._front + 1) % self._cap
        self._count -= 1
        return val

    def get_front(self):
        if self.is_empty():
            raise IndexError("front from empty queue")
        return self._arr[self._front]

    def get_rear(self):
        if self.is_empty():
            raise IndexError("rear from empty queue")
        return self._arr[self._rear]

    def is_empty(self):
        return self._count == 0

    def is_full(self):
        return self._count == self._cap

    def size(self):
        return self._count

    def clear(self):
        self._arr = [None] * self._cap
        self._front = 0
        self._rear = -1
        self._count = 0


if __name__ == "__main__":
    cq = CircularQueue(5)
    for x in [1, 2, 3, 4, 5]:
        cq.enqueue(x)
    print("Front:", cq.get_front(), "Rear:", cq.get_rear())
    try:
        cq.enqueue(6)
    except OverflowError as e:
        print("Expected overflow:", e)
    print("Dequeue ->", cq.dequeue())
    cq.enqueue(6)
    print("After wrap, elements:", [cq.dequeue() for _ in range(cq.size())])
